### DISCUSSION TITLE
Use PiPico as USB HID device

### DISCUSSION THREAD

**Discussion by Sundaresan-JSE-PMS** [micropython/micropython#11077] [RP2040 / Pico]
Hi everyone!

I'm currently working on a project where I need to use my PiPico as a HID (Human Interface Device) device using a USB. So is there any library or code available from Micropython or any other sources? 

Thanks!

**Comment by mattytrentini** [comment]
HID definitions are coming but, for now, you'll have to pick them up from PR #9497 - and see micropython/micropython-lib#558 for an example of how they can be used.

**Comment by dhalbert** [comment]
CircuitPython has mature HID support, including user-defined report descriptors.
https://learn.adafruit.com/circuitpython-essentials/circuitpython-hid-keyboard-and-mouse
https://learn.adafruit.com/customizing-usb-devices-in-circuitpython/hid-devices

**Reply by mattytrentini** [reply]
Yes, and it's solid and easy-to-use!

However, the difference is that MicroPython's goal is to allow USB descriptors to be defined _in Python_. It's proven...challenging to get right!

**Reply by dhalbert** [reply]
Yes, there's a _lot_ of detail. We wrote a library a while ago to "compile" USB descriptors, but didn't get around to doing it for HID Report Descriptors, used it for a while, and eventually gave it up: https://github.com/adafruit/usb_descriptor

**Reply by Sundaresan-JSE-PMS** [reply]
I tried installing Adafruit_Circuitpython_HID library in Thonny, but I'm getting the following error: 
`ERROR: Could not install packages due to an OSError: HTTPConnectionPool(host='127.0.0.1', port=36628): Max retries exceeded with url: /adafruit-blinka/Adafruit_Blinka-8.16.1-py3-none-any.whl/ (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))) Error Command '['C:\\Users\\Rishab\\AppData\\Local\\pipkin\\cache\\workspaces\\6e6d680a4035ac9874f72c843b6111f0\\Scripts\\python.exe', '-I', '-m', 'pip', '--no-color', '--disable-pip-version-check', '--trusted-host', '127.0.0.1', 'install', '--no-compile', '--use-pep517', '--upgrade-strategy', 'only-if-needed', 'adafruit-circuitpython-hid', '--index-url', 'http://127.0.0.1:36628']' returned non-zero exit status 1.`

Any ideas on how I should move forward with this?

**Reply by dhalbert** [reply]
@Sundaresan-JSE-PMS For CircuitPython support, use https://adafru.it/discord or https://forums.adafruit.com/viewforum.php?f=60. No reason to include that stuff here. That appears to be a network problem. I suggest you look at the `circup` tool instead of Thonny for the moment. But let's continue this elsewhere.

**Reply by bobtedbob** [reply]
Circuit python usb already allows you to define USB descriptors in python
